src/pqueue.c

```diff
@@ -0,0 +1,67 @@
+// Implements a min-heap.
+// Modified from http://cs.smu.ca/~pawan/teach/csc342/98-99-2/book/chapter9/minheap.h
+#include <stdlib.h>
+#include <string.h>
+#include "pqueue.h"
+
+void pqueue_init(PQueue *queue, PQueueNode *nodes, size_t num_nodes) {
+  memset(queue, 0, sizeof(*queue));
+
+  queue->nodes = nodes;
+  queue->max_nodes = num_nodes - 1;  // 1-indexed heap - throw away one node
+}
+
+bool pqueue_push(PQueue *queue, void *data, uint16_t prio) {
+  if (queue->size == queue->max_nodes) {
+    // Ran out of space - TODO: handle errors better
```

Comment:
Want me to write a status library that allows for more verbose error handling without exceptions?

ie :
status = 0 = OK
status > 0 = SOME_ENUMERATED_ERROR (optionally with fixed char message).
These could be accompanied by a macro/function that will dump the error to CAN/USB/UART for debugging. Also it allows for some more flexibility with handling multiple failure states from a single function ie if the queue is full then try again after a backoff?